crates/turbopack-ecmascript/src/resolve.rs

```diff
@@ -0,0 +1,121 @@
+use anyhow::Result;
+use turbo_tasks::{Value, Vc};
+use turbopack_core::{
+    issue::{IssueSeverity, IssueSource},
+    reference_type::{CommonJsReferenceSubType, EcmaScriptModulesReferenceSubType, ReferenceType},
+    resolve::{
+        handle_resolve_error,
+        options::ResolveOptions,
+        origin::{ResolveOrigin, ResolveOriginExt},
+        parse::Request,
+        ModuleResolveResult, ModuleResolveResultItem,
+    },
+};
+// TODO: remove
+pub use turbopack_resolve::ecmascript::{apply_cjs_specific_options, apply_esm_specific_options};
+
+use crate::references::external_module::ExternalModuleAsset;
+
+#[turbo_tasks::function]
+pub async fn esm_resolve(
+    origin: Vc<Box<dyn ResolveOrigin>>,
+    request: Vc<Request>,
+    ty: Value<EcmaScriptModulesReferenceSubType>,
+    issue_severity: Vc<IssueSeverity>,
+    issue_source: Option<Vc<IssueSource>>,
+    import_externals: bool,
+) -> Result<Vc<ModuleResolveResult>> {
+    let ty = Value::new(ReferenceType::EcmaScriptModules(ty.into_value()));
+    let options = apply_esm_specific_options(origin.resolve_options(ty.clone()))
+        .resolve()
+        .await?;
+
+    specific_resolve(
+        origin,
+        request,
+        options,
+        ty,
+        issue_severity,
+        issue_source,
+        import_externals,
+    )
+    .await
+}
+
+#[turbo_tasks::function]
+pub async fn cjs_resolve(
+    origin: Vc<Box<dyn ResolveOrigin>>,
+    request: Vc<Request>,
+    issue_source: Option<Vc<IssueSource>>,
+    issue_severity: Vc<IssueSeverity>,
+) -> Result<Vc<ModuleResolveResult>> {
+    // TODO pass CommonJsReferenceSubType
+    let ty = Value::new(ReferenceType::CommonJs(CommonJsReferenceSubType::Undefined));
+    let options = apply_cjs_specific_options(origin.resolve_options(ty.clone()))
+        .resolve()
+        .await?;
+
+    specific_resolve(
+        origin,
+        request,
+        options,
+        ty,
+        issue_severity,
+        issue_source,
+        false,
+    )
+    .await
+}
+
+async fn specific_resolve(
+    origin: Vc<Box<dyn ResolveOrigin>>,
+    request: Vc<Request>,
+    options: Vc<ResolveOptions>,
+    reference_type: Value<ReferenceType>,
+    issue_severity: Vc<IssueSeverity>,
+    issue_source: Option<Vc<IssueSource>>,
+    import_externals: bool,
+) -> Result<Vc<ModuleResolveResult>> {
+    let result = origin.resolve_asset(request, options, reference_type.clone());
+
+    let result = handle_resolve_error(
+        result,
+        reference_type,
+        origin.origin_path(),
+        request,
+        options,
+        issue_severity,
+        issue_source,
+    )
+    .await?;
+
+    replace_externals(result, import_externals).await
```

Comment:
I feel like this behavior might be better placed either in the `ModuleAssetContext` (or in a `ProcessingPlugin`). This way resolving would not depend on the `turbopack-ecmascript` crate and could stay in `turbopack-resolve`, and we could switch it on and off.